src/server/scripts/Kalimdor/zone_durotar.cpp

```diff
@@ -177,9 +178,571 @@ struct npc_mithaka : ScriptedAI
+                        jailer->AI()->SetData(1, 1);
+                    break;
+                case EVENT_TALK_TO_PLAYER:
+                    Talk(SAY_SSEND_YOU_TO_YOUR_DEATH);
+                    me->RemoveUnitFlag(UNIT_FLAG_IMMUNE_TO_PC);
+                    break;
+                case EVENT_DESPAWN_OUT_OF_COMBAT:
+                    me->DespawnOrUnsummon(0s, 2s);
+                    break;
+                case EVENT_CAST_FROSTBOLT:
+                    DoCastVictim(SPELL_FROST_SHOCK);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        if (!UpdateVictim())
+            return;
+
+        DoMeleeAttackIfReady();
+    }
+private:
+    EventMap _events;
+    ObjectGuid _jailorGUID;
+};
+
+enum ClassTrainers
+{
+    EVENT_MOVE_PIT                 = 1,
+    EVENT_MOVE_HOME                = 2,
+
+    GOSSIP_ARCANE_MOMENTUN         = 20690,
+
+    GOSSIP_OPTION_TRAIN_MOMENTUM   = 0,
+    GOSSIP_OPTION_UNTRAIN_MOMENTUM = 1,
+
+    NPC_TRAINER_NORTET             = 38037, // Warrior Trainer
+    NPC_TRAINER_TUNARI             = 38245, // Priest Trainer
+    NPC_TRAINER_SERATHA            = 38246, // Mage Trainer
+    NPC_TRAINER_LEGATI             = 38244, // Rogue Trainer
+    NPC_TRAINER_NEKALI             = 38242, // Shaman Trainer
+    NPC_TRAINER_ERTEZZA            = 38247, // Hunter Trainer
+    NPC_TRAINER_ZENTABRA           = 38243, // Druid Trainer
+    NPC_TRAINER_VOLDREKA           = 42618, // Warlock Trainer
+    NPC_TRAINER_ZABRAX             = 63310, // Monk Trainer
+
+    QUEST_PROVING_PIT_WARRIOR      = 24642,
+    QUEST_PROVING_PIT_PRIEST       = 24786,
+    QUEST_PROVING_PIT_MAGE         = 24754,
+    QUEST_PROVING_PIT_ROGUE        = 24774,
+    QUEST_PROVING_PIT_SHAMAN       = 24762,
+    QUEST_PROVING_PIT_HUNTER       = 24780,
+    QUEST_PROVING_PIT_DRUID        = 24768,
+    QUEST_PROVING_PIT_WARLOCK      = 26276,
+    QUEST_PROVING_PIT_MONK         = 31161,
+
+    SPELL_LEARN_ARCANE_MOMENTUM    = 232062,
+    SPELL_UNLEARN_ARCANE_MOMENTUM  = 232063
+};
+
+// Path point to proving pit fo trainers
+Position const EchoIslandTrainersPitPoints[9] =
+{
+    { -1158.99f, -5421.14f, 13.218976f }, // Nortet Pit
+    { -1137.0f, -5528.23f, 11.979752f },  // Tunari Pit
+    { -1145.95f, -5543.13f, 12.48863f },  // Seratha Pit
+    { -1146.67f, -5430.05f, 13.596256f }, // Legati Pit
+    { -1152.22f, -5407.6f, 13.263395f },  // Nekali Pit
+    { -1136.46f, -5525.13f, 11.99673f },  // Ertezza Pit
+    { -1158.81f, -5533.08f, 11.939185f }, // Zentabra Pit
+    { -1149.92f, -5407.46f, 13.235063f }, // Voldreka Pit
+    { -1151.54f, -5429.86f, 13.29182f }   // Zabrax Pit
+};
+
+// Facing direction at pit for trainers
+float const EchoIslandTrainersPoints[9] = { 0.2094395f, 3.1764990f, 1.7278759f, 1.4835298f, 4.904375f, 3.3161256f, 0.3141593f, 4.956735f, 1.256637f };
+
+template<uint8 PitPos, uint32 QuestID>
+struct npc_echo_isles_class_trainer : public ScriptedAI
```

Comment:
any struct belonging to a given zone_name must have such name in its name

Reply:
done